### PR TITLE
feat: Improves key expiration handling in Explore

### PR DESCRIPTION
**Pull request by michael-s-molina**
### SUMMARY
The objective of this PR is to improve Explore handling when a `form_data_key` expires. Previously when a key expired, the user was presented with the following screen:

Screen Shot 2022-02-08 at 2 08 40 PM

To avoid this situation this PR introduces two new query parameters called `slice_id` and `dataset_id`. These parameters are used when processing an Explore request in case the `form_data_key` is expired. In this scenario, we recover with the minimal information necessary to present the chart, dealing with both saved and unsaved charts.

In the video below I simulate a key expiration by disabling the cache and demonstrate the changes in the workflow:

https://user-images.githubusercontent.com/70410625/153041998-2919104f-e206-4efa-851d-f28d4ff3d641.mov

### TESTING INSTRUCTIONS
Check the video for instructions.

### ADDITIONAL INFORMATION
- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [x] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Review by villebro — COMMENTED** [review]
One comment relating to making the fallback more transparent to the user. But beyond that, really great improvement! 🚀

File: superset/views/core.py [hunk 1]
```diff
@@ -741,6 +741,15 @@ def explore(
...
                 initial_form_data = json.loads(value)
 
+        if not initial_form_data:
+            slice_id = request.args.get("slice_id")
+            if slice_id:
+                initial_form_data["slice_id"] = slice_id
+
+            dataset_id = request.args.get("dataset_id")
+            if dataset_id:
+                initial_form_data["datasource"] = f"{dataset_id}__table"
```
Comment (zhaoyongjie) [inline superset/views/core.py:751]: simplify codes

```
        slice_id = request.args.get("slice_id")
        dataset_id = request.args.get("dataset_id")
        if not initial_form_data and not slice_id and not dataset_id:
            flash('some error message if can't get slice_id and dataset_id')

        if not initial_form_data:
            initial_form_data["slice_id"] = slice_id
            initial_form_data["datasource"] = f"{dataset_id}__table"
```

File: superset-frontend/src/dashboard/components/menu/ShareMenuItems/ShareMenuItems.test.tsx [hunk 2]
```diff
@@ -159,7 +159,7 @@ test('Click on "Share dashboard by email" and succeed', async () => {
   await waitFor(() => {
     expect(props.addDangerToast).toBeCalledTimes(0);
     expect(window.location.href).toBe(
-      'mailto:?Subject=Superset dashboard COVID Vaccine Dashboard%20&Body=Check out this dashboard: http://localhost:8088/r/3',
+      'mailto:?Subject=Superset dashboard COVID Vaccine Dashboard%20&Body=Check out this dashboard: http%3A%2F%2Flocalhost%3A8088%2Fr%2F3',
```
[the remark this answers is not in the document]
Comment (michael-s-molina) [reply superset-frontend/src/dashboard/components/menu/ShareMenuItems/ShareMenuItems.test.tsx]: Nice catch!

File: superset/views/core.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (michael-s-molina) [reply superset/views/core.py]: Good idea. Done.

Comment (michael-s-molina) [reply superset/views/core.py]: I changed the logic a little bit to only use one or the other and present a specific message for each type 😉

**Review by zhaoyongjie — APPROVED** [review]
Thanks for the fix. LGTM

**Review by villebro — APPROVED** [review]
LGTM!

**Comment by ktmud** [comment]
I'm wondering if we can encode the slice and dataset ids directly in the URL path. I.e., the explore URL becomes `/explore/s/{slice_id}/{form_data_key}` or `/explore/d/{dataset_id}/{form_data_key}`

**Pull request merged**